### DISCUSSION TITLE
An @conditional that always parses its body

### DISCUSSION THREAD

**Discussion by vpax** [zeek/zeek#2966] [Ideas]
First let me describe the problems I'm trying to solve:

1. `-O gen-C++` currently has to skip functions (or in some cases whole files) that are influenced by `@if` conditionals, because if it compiles them in one context and they execute in another (where the `@if` would resolve differently), the results would be incorrect. There's no easy fix for this because `@if`s throw away their enclosed body un-scanned/un-parsed when the conditional is false.
2. I want to develop extensive scripting code-coverage evaluations, so I can spot where our BTests have holes. One such hole is conditional code that our test suite fails to exercise. Currently, I could tell that there are _lines_ in the script that aren't getting covered, but not the specifics (like the names of functions defined in those lines).

To address these, I'm thinking about introducing a new type of `@if` conditional, say `@activate-if`.  The key differences from `@if` would be (1) the expression in the conditional is remembered, and made available to subsequent analysis (such as allowing the compile-to-C++ compiler to generate run-time conditional decisions), (2) code inside would be (a) fully parsed (resulting in errors if malformed, or if this introduces inconsistencies with other code outside the conditional), but (b) _deactivated_ (event handlers wouldn't be installed, `redef`s wouldn't happen), and (3) perhaps some forms of scripting would be disallowed or generate warnings, such as extending a record, or complaining about appearing inside a non-activating `@if`, or (maybe) if a `@load` is present.

Not all `@if`s can be treated in this way, which is why this would be a new form of conditional rather than an adjustment to the current one.  In current use, `@if`s are heavily dominated by varying behavior depending on whether the script is running on a cluster, and if so whether the current node is a worker/proxy/logger.  From the inspections I've done so far, it looks like many of these could be converted to `@activate-if`s.

**Comment by awelzel** [comment]
> and (3) perhaps some forms of scripting would be disallowed or generate warnings, such as extending a record, or complaining about appearing inside a non-activating @if, or (maybe) if a @load is present.

Following an `@load` inside a non-activated context bid us due to the activated context not doing the load anymore:
https://github.com/zeek/zeek/issues/3098

This also brings up the question how plugins should deal with this because `@load` can be handled by plugins via `HookLoadFile()`.

Not following `@load` might be easiest, but also rather limiting in coverage.

**Comment by vpax** [comment]
@awelzel do you think there are any prospects for reviving this? I'm now doing a big push on test-coverage for our content and the lack of functionality lack this continues to concern me since it makes for blind spots that are very difficult to otherwise spot.

One "bandaid" thought would be a different mechanism, one that allows for adding conditionals to Zeek's code coverage output, such that if in a test suite run a given conditional never is active, that would be apparent. That might suffice for my purposes, rather than getting line-level granularity for code that isn't exercised due to being behind a conditional.

**Reply by awelzel** [reply]
> @awelzel do you think there are any prospects for reviving this?

I'm pretty neutral. We reverted the original version primarily due to timing as we were about to cut a release. It was also a fairly sizable change, including both, replacing `@if` with runtime `if`, but also the `@if-analyze` infrastructure IIRC.

I'd be more than happy if we'd move to less `@if`-based cluster code, but probably in individual PRs independent of `@if-analyze` or other changes. 

> One "bandaid" thought would be a different mechanism, one that allows for adding conditionals to Zeek's code coverage output, such that if in a test suite run a given conditional never is active, that would be apparent.

I haven't used the Zeek script coverage functionality at all and I'm not sure we have good docs on it, but if in the end there'd be a HTML page (or some text based output) that would show the lines between some `@if`  and `@endif` pair as never parsed, that seems pretty useful. I guess that's then not "script execution" coverage anymore, but "parse coverage" or so. It doesn't feel like that needs a lot of infrastructure if one limits it to be just line-based?

I do have some  other thoughts, loosely related to this topic, mostly with focus on cluster functionality:

* Within base Zeek, we've added `test-all-policy-cluster.test` that loads all scripts as manager, logger, proxy and worker node to verify they load properly. I wonder if that sort of test could be generalized for package content - doesn't need to be in Zeek base, maybe a generic btest added to all packages that you maintain. A bit ad-hoc, but also seems pragmatic and somewhat easy to explain what the test does.

* Without any numbers to back this, I'm assuming that `@if` guards are mostly about cluster functionality, so:
  * With the deprecation of `Broker::auto_publish()`, does using explicit `Cluster::publish()` lead to more `if` runtime guards, making this less of an issue down the road?
  * Totally out there: I'm still sometimes thinking of a "loopback" or "local" cluster backend, such that in single-node mode, events published via `Cluster::publish()` are received back locally, it's a very different model to what there is today, but it could be powerful.

* `Broker::publish()` / `Cluster::publish()` the `make_event()` bifs are variable args, so you're still prone to run into runtime errors if one gets the arguments wrong. Far in the back of my mind, I wanted to attempt writing a AST traverser that inspects `Cluster::publish()` and `Cluster::make_event()` statically at InitPostScript() time. I.e, take the event arg and passed args and verify statically that their types match.
  ```
  event ping(c: count) { }
  # Mistakenly pass a string instead of count
  Cluster::publish("/topic", ping, "42");
  ```
  Today, static parsing says this is fine, but at runtime there's an error.
  ```
  $ zeek --parse-only ping.zeek 
  $ zeek ping.zeek 
  error in ./ping.zeek, line 5: event parameter #1 type mismatch, got string, expecting count
  ```
  While this is ad-hoc for `Cluster::publish()` or `Cluster::make_event()`, to me parsing scripts for each cluster node *and* having this functionality might provide more bang for the buck. Pondering, I think we should have that static check one way or another if it has a chance to capture that `ping("42")` error above.


TLDR: I wonder if with your "parse coverage" idea, a generic load-as-all-node-types-test, and also static type checking for `Cluster::publish(...)` and `Cluster:::make_event()` calls, we get into a good enough place?

**Comment by vpax** [comment]
A few thoughts:

- If code-coverage included a notion of "was parsed" (i.e., not skipped), that could suffice for my purposes, but it'd be a bit involved because I'd need to mine out all of the conditionals in the scripts being assessed to then compare against that list. Easier on my end would be if code-coverage simply reported per-conditional whether it was True or False during some execution. I could then look for conditionals that didn't have coverage spanning both True and False to find areas lacking coverage.
- It's not just cluster stuff (though that's the most common) but also conditionals that inspect the Zeek version, because we generally have to support both the last LTS and the upcoming one (i.e., we try to track Zeek main), and deal with deprecation differences.
- In our scripts there are also conditionals that test whether another package has been loaded. (This actually points up a shortcoming of `zkg`, which is it would be darn handy if a package A could indicate that it _cares_ about another package B, but doesn't _depend_ on it. That would tell `zkg` that _if_ B is going to be loaded, then that load should come before the load of A.)
- Regarding something like `test-all-policy-cluster.test`, I don't think a generic form would suffice to fully exercise the clusterized content. It would though catch parse errors, which is already handy.
- Re type-checking for variadic BiFs - _yes!_ This would be great to add. You can imagine doing it via an extension to `bifcl` that would allow a BiF developer to specify a predicate that assesses whether a given call type-checks. The Zeek parser would then call this per BiF instantiation.
